src/librustc/ty/mod.rs

```diff
@@ -1483,10 +1485,29 @@ impl<'gcx> HashStable<StableHashingContext<'gcx>> for AdtDef {
             ref repr,
         } = *self;
 
-        did.hash_stable(hcx, hasher);
-        variants.hash_stable(hcx, hasher);
-        flags.hash_stable(hcx, hasher);
-        repr.hash_stable(hcx, hasher);
+        thread_local! {
+            static CACHE: RefCell<FxHashMap<usize, Fingerprint>> =
+                RefCell::new(FxHashMap());
+        }
+
+        let hash = CACHE.with(|cache| {
+            let addr = self as *const AdtDef as usize;
+            if let Some(&hash) = cache.borrow().get(&addr) {
```

Comment:
Since `AdtDef`s are not recursive, you should be able to slightly optimize this by using `HashMap::entry()`. That avoids having to look up the key twice in the "cache miss" case. 

You cannot do the same for `ty::Slice` because while hashing one slice, you might encounter a nested one somewhere down the line and consequently run into a `RefCell` double-borrow panic.